### DISCUSSION TITLE
Reading all control state

### DISCUSSION THREAD

**Discussion by ahopper** [softerhardware/CWKeyer#11] [Ideas]
To create a ui able to set all possible settings you really need access to the state of all controls, not just the ones that have changed since connection.
This could just be a single message to trigger a send of all values or some way to request each in turn.

**Comment by recri** [comment]
Hasak sends its settings as they're modified, so it produces a complete
inventory when it powers up and sets the initial defaults.  Those may come
so fast that they trample each other.

It will also produce a throttled roll call any time it receives NRPN 2003.

As soon as I can get the quotation marks right, there will be hasak.json to
go with every released hasak.ino listing all the NRPNs and their details.

-- 73 -- rec -- ad5dz --

**Reply by dl1ycf** [reply]
"Hasak sends its settings as they're modified, so it produces a complete
inventory when it powers up and sets the initial defaults."

The problem is, that the Teensy powers up as soon as the computer is switched on, and any "client" listening to those messages is started much later (after the computer boot is completed). So everything sent from the Teensy upon power-up is easily lost.

**Comment by ahopper** [comment]
That sounds perfect, being able to demand them is good as the sdr sw is probably not running when the keyer starts up.
I'm making the keyer ui in spark driven from a .json file so I shall try and align it with yours.  
The icing on the cake would be to get an equivalent of the .json file from the keyer.

**Reply by recri** [reply]
I guess I'm unclear, you want the descriptor JSON, or you want a JSON formatted dump of the current settings.

I could compile the JSON descriptor file into character array in the firmware and then spool it out to you over MIDI.

Then you wouldn't need to figure out what version you were talking to, it would describe itself.

**Reply by ahopper** [reply]
Yes I meant the json descriptor.  This would then allow anyone to add features and the ui would keep up without having to coordinate version numbers and locations of files.

**Comment by recri** [comment]
That may be a problem for some people, but I just unplug the Teensy from
USB and plug it back in.  It starts, it enumerates its settings, it's
done.  But I haven't needed it recently.

**Comment by SimonG4ELI** [comment]
Windows:

 

*	I have a good json class in a single header (from Github) should anyone need it.
*	As we’re talking CW latency in the RX chain is important. I’ve just implemented Partitioned FFT Convolution (IF filtering) which has a latency of 4ms irrespective of filter size. Will give links etc. should anyone be interested.

 

Simon Brown, G4ELI

 <https://www.sdr-radio.com/> https://www.sdr-radio.com

**Reply by dl1ycf** [reply]
No this is a misunderstanding. Latency in the RX chain is of no concern when it comes to keying, this is all TX.

**Comment by SimonG4ELI** [comment]
If you're using a SDR for CW (or SSB for that matter) you don't want unnecessary RX latency as someone else can jump in before you in reply to a CQ. Ideally every area should have the lowest latency possible. Using 'Partitioned FFT Convolution' in the RX is a good way of reducing delay in the RX filtering. When you're in a QSK QSO at 50wpm then latency counts.

FWIW the IC-7300 latency is excellent.

**Reply by dl1ycf** [reply]
While this is all true, it has nothing to do with keying.  What counts here is that the side tone is synchronous with the internal state of the keyer. BTW in Linux and (a little bit less so) in MacOS, a large part of the RX latency results from what happens with the RX audio samples *after* the filtering, when they are "sent" to ALSA, pulseaudio, portaudio or whatever audio subsystem in use. Here we are talking about 50 msec or so in bad cases.

**Reply by SimonG4ELI** [reply]
On Windows (and I believe Linux) you can bypass the mixer and talk direct to the audio device driver, bringing latency down to sub-10ms.

**Reply by dl1ycf** [reply]
So is there an experiment with a two-channel oscilloscope showing the RF input and the headphone output in two traces?
The topic is really important (although a little off-topic here) and this kind of measurement is what counts.

In my experience the question is not how low you *can* go in terms of audio latency, but how low you *want* to go, to be on the safe side avoiding audio crackles if, e.g. a HPSDR packet from the radio arrives somewhat late.

**Reply by ahopper** [reply]
Hi, not sure how this message ended up in this thread, using the github interface is recommended.
Simon, yep Partitioned fft convolution is good and is what spark used from day one (5+ yrs ago)  and anything using wdsp used shortly after. There are a number of other tricks to reduce the various latencies in sdr, there is a thread somewhere on the Hermes lite group of my early experiments on midi cw which indicates far lower figures are possible than many that have come up recently (using the twin scope trace measurement). I shall re run with the current setup, but the key thing is this keyer is designed to solve one part of the challenge and does so very well.

**Comment by recri** [comment]
Okay, that makes excellent sense to me.

The discussion "Expanded NRPN common core #23" includes a NRPN_ID_JSON
nrpn, please chime in.  And help me find any bugs in the NRPN_STRING*
proposal.
